Controllers/Api/Dashboard/PowerbiReferenceController.cs

```diff
@@ -206,5 +289,271 @@ public async Task<IActionResult> Delete(int id)
 			return this.BadRequest(logUnknown);
 
 		}
+
+		[HttpPut("share/{powerbiId}")]
+		public async Task<IActionResult> Share([FromBody] List<UserShare> shareTos, int powerbiId)
+		{
+			var activity = new ApplicationActivity() { Controller = ControllerName, Action = $"Share/{powerbiId}" };
+
+			var owner = this.User.Identity.Name;
+
+			try
+			{
+				var hasPermission = await this.accessProfileHelper.HasPermission(owner, FeatureNames.Dashboard, Security.AccessLevelFlag.Write);
+				if (!hasPermission)
+				{
+					var log = await logger.LogInvalidAccess(activity: activity, relevantObject: nameof(PowerbiReference), log: $"Insufficient access");
+					return this.BadRequest("Insufficient access");
+				}
+
+				#region ServerSide Validation
+				if (!this.ModelState.IsValid)
+				{
+					var log = await logger.LogInvalidData(activity, "Invalid model");
+					return this.BadRequest(log);
+				}
+				#endregion ServerSide Validation
+
+				var powerbiReference = await this.dataRepository.GetPowerbiReferenceById(powerbiId);
+				if (powerbiReference == null)
+				{
+					var log = await logger.LogFailure(activity: activity, relevantObject: nameof(PowerbiReference), reference: $"{powerbiId}", log: $"Not found");
+					return this.BadRequest(log);
+				}
+
+				if (powerbiReference.Sharing?.UserShares == null)
+				{
+					powerbiReference.Sharing = new Sharing()
+					{
+						UserShares = shareTos
+					};
+				}
```

Comment:
I believe that the condition 'powerbiReference.Sharing?.UserShares == null' will never be true as long as Sharing itself is not null since we initialize the UserShares to be empty list by default 'public ICollection<UserShare> UserShares { get; set; } = new List<UserShare>();'

```suggestion
				if (powerbiReference.Sharing == null)
				{
					powerbiReference.Sharing = new Sharing()
					{
						UserShares = shareTos
					};
				}
```